File: src/main/java/com/wxl/dyttcrawler/config/CrawlerConfiguration.java
package com.wxl.dyttcrawler.config;

import com.wxl.dyttcrawler.core.*;
import com.wxl.dyttcrawler.processor.DyttProcessor;
import com.wxl.dyttcrawler.processor.ProcessorDispatcher;
import com.wxl.dyttcrawler.properties.CrawlerProperties;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.ObjectProvider;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;
import org.springframework.core.env.Environment;
import org.springframework.util.CollectionUtils;
import us.codecraft.webmagic.Site;
import us.codecraft.webmagic.downloader.Downloader;
import us.codecraft.webmagic.pipeline.Pipeline;
import us.codecraft.webmagic.processor.PageProcessor;
import us.codecraft.webmagic.scheduler.Scheduler;

import java.util.HashSet;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Create by wuxingle on 2020/5/10
 * 爬虫配置
 */
@Slf4j
@Configuration
@EnableConfigurationProperties(CrawlerProperties.class)
public class CrawlerConfiguration {

    private CrawlerProperties crawlerProperties;

    private Environment environment;

    @Autowired
    public CrawlerConfiguration(CrawlerProperties crawlerProperties,
                                Environment environment) {
        this.crawlerProperties = crawlerProperties;
        this.environment = environment;
    }

    /**
     * 爬虫管理
     */
    @Bean(destroyMethod = "close")
    public Crawler crawler(PageProcessor pageProcessor,
                           Downloader downloader,
                           ObjectProvider<Pipeline> pipelines,
                           ThreadPool threadPool,
                           Scheduler scheduler,
                           ObjectProvider<CrawlerListener> listeners) {
        String taskId = environment.getRequiredProperty("spring.application.name");
        String firstUrl = crawlerProperties.getFirstUrl();

        return Crawlers.create()
                .setTaskId(taskId)
                .setPageProcessor(pageProcessor)
                .setDownloader(downloader)
                .addPipeline(pipelines.orderedStream().collect(Collectors.toList()))
                .setScheduler(scheduler)
                .setThreadPool(threadPool)
                .startUrls(firstUrl)
                .addCrawlerListeners(listeners.orderedStream().collect(Collectors.toList()))
                .setExitWhenComplete(true)
                .build();
    }

    /**
     * 线程池
     */
    @Bean
    public ThreadPool crawlerExecutorService() {
        return new ExecutorThreadPool(crawlerProperties.getMaxThreads());
    }

    /**
     * 页面处理
     */
    @Bean
    public ProcessorDispatcher processorDispatcher(Site site, ObjectProvider<DyttProcessor> processors) {
        List<DyttProcessor> collect = processors.orderedStream().collect(Collectors.toList());
        return new ProcessorDispatcher(site, collect);
    }

    /**
     * site配置
     */
    @Primary
    @Bean("dyttPrimarySite")
    public Site dyttPrimarySite() {
        CrawlerProperties.SiteProperties siteProp = crawlerProperties.getSite();
        Site site = Site.me()
                .setCharset(crawlerProperties.getCharset())
                .setUserAgent(siteProp.getUserAgent())
                .setSleepTime((int) siteProp.getSleepTime().toMillis())
                .setRetryTimes(siteProp.getRetryTimes())
                .setRetrySleepTime((int) siteProp.getRetrySleepTime().toMillis())
                .setTimeOut((int) siteProp.getTimeout().toMillis())
                .setDisableCookieManagement(siteProp.isDisableCookie());

        if (!CollectionUtils.isEmpty(siteProp.getAcceptStatusCode())) {
            site.setAcceptStatCode(new HashSet<>(siteProp.getAcceptStatusCode()));
        }
        if (!CollectionUtils.isEmpty(siteProp.getHeaders())) {
            siteProp.getHeaders().forEach(site::addHeader);
        }
        return site;
    }

}

File: src/main/kotlin/com/wxl/dyttcrawler/config/CrawlerConfiguration.kt
package com.wxl.dyttcrawler.config

import com.google.common.util.concurrent.ThreadFactoryBuilder
import com.wxl.dyttcrawler.core.Crawler
import com.wxl.dyttcrawler.core.CrawlerCoroutineScope
import com.wxl.dyttcrawler.core.CrawlerListener
import com.wxl.dyttcrawler.downloader.HttpDownloader
import com.wxl.dyttcrawler.processor.DyttProcessor
import com.wxl.dyttcrawler.processor.ProcessorDispatcher
import com.wxl.dyttcrawler.properties.CrawlerProperties
import com.wxl.dyttcrawler.properties.SiteProperties
import kotlinx.coroutines.asCoroutineDispatcher
import org.springframework.beans.factory.ObjectProvider
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.env.Environment
import us.codecraft.webmagic.Request
import us.codecraft.webmagic.Site
import us.codecraft.webmagic.pipeline.Pipeline
import us.codecraft.webmagic.processor.PageProcessor
import us.codecraft.webmagic.scheduler.Scheduler
import java.util.concurrent.SynchronousQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit
import java.util.stream.Collectors

/**
 * Create by wuxingle on 2021/10/02
 * 爬虫配置
 */
@Configuration
class CrawlerConfiguration(
    private val crawlerProperties: CrawlerProperties,
    private val siteProperties: SiteProperties,
    private val environment: Environment
) {

    /**
     * 爬虫管理
     */
    @Bean(destroyMethod = "close")
    fun crawler(
        pageProcessor: PageProcessor,
        downloader: HttpDownloader,
        pipelines: ObjectProvider<Pipeline>,
        coroutineScope: CrawlerCoroutineScope,
        scheduler: Scheduler,
        listeners: ObjectProvider<CrawlerListener>
    ): Crawler? {
        val taskId = environment.getRequiredProperty("spring.application.name")
        val startUrl = crawlerProperties.startUrl
        val concurrentNum = crawlerProperties.concurrentNum

        return Crawler.build {
            this.taskId = taskId
            this.pageProcessor = pageProcessor
            this.downloader = downloader
            this.pipelines = pipelines.orderedStream().collect(Collectors.toList())
            this.scheduler = scheduler
            this.coroutineScope = coroutineScope
            this.startRequests = listOf(Request(startUrl))
            this.exitWhenComplete = true
            this.concurrentNums = concurrentNum
            this.crawlerListeners = listeners.orderedStream().collect(Collectors.toList())
        }
    }

    /**
     * 爬虫线程池
     */
    @Bean(destroyMethod = "shutdown")
    fun crawlerExecutorService(): ThreadPoolExecutor {
        val maxThreads = crawlerProperties.maxThreads
        return ThreadPoolExecutor(
            1, maxThreads,
            10, TimeUnit.MINUTES,
            SynchronousQueue(),
            ThreadFactoryBuilder().setDaemon(true).setNameFormat("crawler-pool-%s").build(),
            ThreadPoolExecutor.AbortPolicy()
        )
    }

    /**
     * 爬虫的CoroutineScope
     */
    @Bean
    fun crawlerCoroutineScope(poolExecutor: ThreadPoolExecutor): CrawlerCoroutineScope {
        val dispatcher = poolExecutor.asCoroutineDispatcher()
        return CrawlerCoroutineScope(dispatcher)
    }

    /**
     * 页面处理
     */
    @Bean
    fun processorDispatcher(site: Site, processors: ObjectProvider<DyttProcessor>): ProcessorDispatcher {
        val collect: List<DyttProcessor> = processors.orderedStream().collect(Collectors.toList())
        return ProcessorDispatcher(site, collect)
    }

    /**
     * site配置
     */
    @Bean
    fun dyttPrimarySite(): Site = Site.me().apply {
        domain = siteProperties.domain
        charset = siteProperties.charset
        userAgent = siteProperties.userAgent
        sleepTime = siteProperties.sleepTime.toMillis().toInt()
        retryTimes = siteProperties.retryTimes
        timeOut = siteProperties.timeout.toMillis().toInt()
        if (siteProperties.acceptStatusCode.isNotEmpty()) {
            acceptStatCode = siteProperties.acceptStatusCode.toSet()
        }
        if (siteProperties.headers.isNotEmpty()) {
            for ((k, v) in siteProperties.headers) {
                addHeader(k, v)
            }
        }
        isUseGzip = siteProperties.useGzip
        isDisableCookieManagement = true
    }

}